File: 103-BinaryTreeZigzagLevelOrderTraversal.cpp
/*

Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).

For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its zigzag level order traversal as:
[
  [3],
  [20,9],
  [15,7]
]

*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root) return {};
        queue<TreeNode *> q;
        q.push(root);
        int count = 1;
        vector<vector<int>> res;
        vector<int> curr;
        bool straight = false;
        
        while(!q.empty()) {
            TreeNode* f = q.front();
            q.pop();
            curr.push_back(f->val);
            -- count;
            if(f->left) q.push(f->left);
            if(f->right) q.push(f->right);
            if (count == 0) {
                if(straight) {
                    reverse(curr.begin(), curr.end());
                }
                res.push_back(curr);
                curr.clear();
                count = q.size();
                straight = !straight;
            }
        }
        return res;
    }
};

File: java/103-BinaryTreeZigzagLevelOrderTraversal.java
/*

Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).

For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its zigzag level order traversal as:
[
  [3],
  [20,9],
  [15,7]
]

*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if(root == null) return res;
        Queue<TreeNode> q = new LinkedList<>();
        List<Integer> curr = new ArrayList<>();
        q.offer(root);
        boolean flag = true;
        int n = 1;
        while(!q.isEmpty()) {
            System.out.println("1");
            TreeNode t = q.poll();
            --n;
            curr.add(t.val);
            if(t.left != null) q.offer(t.left);
            if(t.right != null) q.offer(t.right);
            if(n == 0) { //level complete
                if(flag == true) {
                    res.add(new ArrayList<>(curr));
                }
                else {
                    List<Integer> temp = new ArrayList<>(curr);
                    Collections.reverse(temp);
                    res.add(temp);
                }
                flag = !flag;
                curr.clear();
                n = q.size();
            }
        }
        return res;         
    }
}
